services/controlplane/features/client_test.go

```diff
@@ -0,0 +1,118 @@
+package features_test
+
+import (
+	"context"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/rudderlabs/rudder-server/services/controlplane/features"
+	"github.com/rudderlabs/rudder-server/services/controlplane/identity"
+	"github.com/stretchr/testify/require"
+)
+
+func Test_Client_Send(t *testing.T) {
+	validSecret := "valid-secret"
+
+	wantBody := `{"components":[{"name":"test","features":["feature1","feature2"]}]}`
+
+	testcases := []struct {
+		name     string
+		identity identity.Identifier
+
+		wantPath string
+		wantErr  error
+	}{
+		{
+			name: "workspace",
+			identity: &identity.Workspace{
+				WorkspaceID:    "valid-workspace-id",
+				WorkspaceToken: validSecret,
+			},
+
+			wantPath: "/data-plane/v1/workspaces/valid-workspace-id/settings",
+			wantErr:  nil,
+		},
+		{
+			name: "namespace",
+			identity: &identity.Namespace{
+				Namespace:    "valid-namespace",
+				HostedSecret: validSecret,
+			},
+			wantPath: "/data-plane/v1/namespaces/valid-namespace/settings",
+			wantErr:  nil,
+		},
+	}
+
+	for _, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			s := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				u, _, _ := r.BasicAuth()
+
+				require.Equal(t, "application/json", r.Header.Get("Content-Type"))
+				require.Contains(t, r.UserAgent(), "control-plane/features")
+				require.Equal(t, validSecret, u)
+				require.Equal(t, http.MethodPost, r.Method)
+				require.Equal(t, tc.wantPath, r.URL.Path)
+
+				body, err := io.ReadAll(r.Body)
+				require.NoError(t, err)
+				require.Equal(t, wantBody, string(body))
+
+				w.WriteHeader(http.StatusNoContent)
+			}))
+			defer s.Close()
+
+			c := features.New(tc.identity, features.WithHTTPClient(s.Client()), features.WithURL(s.URL))
+
+			err := c.Send(context.Background(), features.PerComponent{
+				"test": {"feature1", "feature2"},
+			})
+			require.NoError(t, err)
+		})
+	}
+
+	t.Run("unexpected status", func(t *testing.T) {
+		count := int64(0)
+		s := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			atomic.AddInt64(&count, 1)
+			w.WriteHeader(http.StatusBadRequest)
+		}))
+		defer s.Close()
+
+		c := features.New(&identity.Namespace{}, features.WithHTTPClient(s.Client()), features.WithURL(s.URL))
+
+		err := c.Send(context.Background(), features.PerComponent{
+			"test": {"feature1", "feature2"},
+		})
+		require.Error(t, err)
+
+		require.Equalf(t, int64(4), atomic.LoadInt64(&count), "retry %d times", features.MaxRetries)
+	})
+
+	t.Run("timeout", func(t *testing.T) {
+		t.Skip("test is too slow")
```

Comment:
Why having a test if we skip it? Also, given that you set the timeout to one millisecond, why is the test slow? With the backoff included we are below 3secs. Additionally we could inject the retry strategy in the `*Client` and have it default to the backoff one. WDYT?

Reply:
If you do something like [this](https://github.com/rudderlabs/rudder-server/commit/c67635452140ffb0391c9c5a47f97a72ea9da7f8) the test goes down to 0.01s of execution plus we get a new feature :+1:

Reply:
Good point; I've overridden the retries to one. It solves the problem without introducing coupling. retries are tested with a different test either way.